detectorSegmentations/src/GridDRcalo_k4geo.cpp

```diff
@@ -0,0 +1,238 @@
+#include "detectorSegmentations/GridDRcalo_k4geo.h"
+
+#include <climits>
+#include <cmath>
+#include <stdexcept>
+
+namespace dd4hep {
+namespace DDSegmentation {
+
+/// default constructor using an encoding string
+GridDRcalo_k4geo::GridDRcalo_k4geo(const std::string& cellEncoding) : Segmentation(cellEncoding) {
+  // define type and description
+  _type = "GridDRcalo_k4geo";
+  _description = "DRcalo segmentation based on the tower / (Cerenkov or Scintillation) fiber / SiPM hierarchy";
+
+  // register all necessary parameters
+  registerIdentifier("identifier_assembly", "Cell ID identifier for assembly", fAssemblyId, "assembly");
+  registerIdentifier("identifier_eta", "Cell ID identifier for numEta", fNumEtaId, "eta");
+  registerIdentifier("identifier_phi", "Cell ID identifier for numPhi", fNumPhiId, "phi");
+  registerIdentifier("identifier_x", "Cell ID identifier for x", fXId, "x");
+  registerIdentifier("identifier_y", "Cell ID identifier for y", fYId, "y");
+  registerIdentifier("identifier_IsCerenkov", "Cell ID identifier for IsCerenkov", fIsCerenkovId, "c");
+  registerIdentifier("identifier_module", "Cell ID identifier for module", fModule, "module");
+
+  fParamBarrel = new DRparamBarrel_k4geo();
+  fParamEndcap = new DRparamEndcap_k4geo();
+}
+
+GridDRcalo_k4geo::GridDRcalo_k4geo(const BitFieldCoder* decoder) : Segmentation(decoder) {
+  // define type and description
+  _type = "GridDRcalo_k4geo";
+  _description = "DRcalo segmentation based on the tower / (Cerenkov or Scintillation) fiber / SiPM hierarchy";
+
+  // register all necessary parameters
+  registerIdentifier("identifier_assembly", "Cell ID identifier for assembly", fAssemblyId, "assembly");
+  registerIdentifier("identifier_eta", "Cell ID identifier for numEta", fNumEtaId, "eta");
+  registerIdentifier("identifier_phi", "Cell ID identifier for numPhi", fNumPhiId, "phi");
+  registerIdentifier("identifier_x", "Cell ID identifier for x", fXId, "x");
+  registerIdentifier("identifier_y", "Cell ID identifier for y", fYId, "y");
+  registerIdentifier("identifier_IsCerenkov", "Cell ID identifier for IsCerenkov", fIsCerenkovId, "c");
+  registerIdentifier("identifier_module", "Cell ID identifier for module", fModule, "module");
+
+  fParamBarrel = new DRparamBarrel_k4geo();
+  fParamEndcap = new DRparamEndcap_k4geo();
+}
+
+GridDRcalo_k4geo::~GridDRcalo_k4geo() {
+  if (fParamBarrel) delete fParamBarrel;
```

Comment:
`delete nullptr` is essentially a no-op, so the check for nullptr-ness is unnecessary.
What might be useful is setting, e.g.,  `fParamBarrel=nulltpr` in the constructor to prevent the attempt from `delete`ing random memory.